Add BSpline Utility

lipohlmann opened the pull request:
closes #30820

## Reason
Utility is needed for continued work on a mesh generator to connect objects.

## Design
Implements an arbitrary BSpline recursive formula using a given vector of control points and a desired interpolation degree. Intended to be extremely flexible.

## Impact
Allows for continued work on the mesh generator project described above while also providing a tool found in many engineering applications. Utility will likely find use in others' projects.

loganharbour committed:
Add basic structure for BSpline utility

Commit:
Create BSpline utility.
Code takes in a vector of libMesh Points as control points and an interpolating polynomial degree and returns a BSpline object,
which can be evaluated for values of t on the standard interval [0,1].
Unit test cases were created to verify degree-varying capabilities and user feedback when the number of control points is insufficient.

closes #30820

roystgnr commented on framework/src/utils/BSpline.C:
```diff
@@ -0,0 +1,118 @@
+  const unsigned int num_points_left = _control_points.size() - _degree;
+
+  if (num_points_left <= 0)
+    mooseError("Number of control points must be equal to degree + 1!");
```
This should be "greater than or equal to", no?

roystgnr replied:
Or just "greater than polynomial degree"?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2978479/), step Clang format on 63da9e5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30845/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30845/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b99593cafbe0b2f62f23e00ce7b5f69f05d493c6`

Commit:
Address greater than or equal to inconsistency in comments for buildKnotVector method.

Commit:
remove namespace from header file, add mooseAssert lines, adjust comments in source and header file

roystgnr commented:
This looks good to me.  Since we're not in any hurry to merge before we have the use case ready, though, let's hold off to give others time to look at it.

Honestly, when I recommended that you do the utility PR first separate from the generator PR, I was kind of expecting that the utility PR would have more about it to criticize and fix.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2979280/), step Docs: sync website on 670e807 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30845/site)

This comment will be updated on new commits.

roystgnr commented on framework/include/utils/BSpline.h:
```diff
@@ -0,0 +1,57 @@
+#include "libmesh/point.h"
+#include "libMeshReducedNamespace.h"
+
+using namespace libMesh;
```
Oh, man - did I miss it when you got this out of Moose.h last year?  I thought that was going to be stuck there forever.  Good job!

LP1012 replied:
To be clear, is it the `using namespace libMesh` that is the issue? Or are the other includes a problem?

Commit:
Address review:
- add docstrings
- add assert on polynomial degree

GiudGiud commented on framework/src/utils/BSpline.C:
```diff
@@ -0,0 +1,118 @@
+ */
+libMesh::Real
+BSpline::firstCoeff(const libMesh::Real & t, const unsigned int i, const unsigned int j) const
+{
```
```suggestion
{
  mooseAssert(i + j < _knot_vector.size(), "Index must stay within bounds");
```

GiudGiud commented on framework/src/utils/BSpline.C:
```diff
@@ -0,0 +1,118 @@
+libMesh::Real
+BSpline::firstCoeff(const libMesh::Real & t, const unsigned int i, const unsigned int j) const
+{
+  libMesh::Real ti = _knot_vector[i];
+  libMesh::Real tij = _knot_vector[i + j];
```
```suggestion
  const auto ti = _knot_vector[i];
  const auto tij = _knot_vector[i + j];
```

lipohlmann committed:
Address Review
Add MooseAssert on knot vector access

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Commit:
Fix knot_vector submethod

fix clang formatting

Remove useless assert
Add reference paper to header

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2998186/), step Clang format on f43600f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30845/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30845/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8056243abde862a63e2d0d54390ce3b2c6eafe84`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2998449/), step Generate coverage on c1fcb4a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>805624</th>
      <th colspan=4>#30845 c1fcb4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.47%</td>
      <td>+0.00%</td>
      <td>97.56%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>114869</td>
      <td>114907</td>
      <td>+38</td>
      <td>40</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19524</td>
      <td>19527</td>
      <td>+3</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30845/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30845/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30845/coverage/xfem)

</details>



This comment will be updated on new commits.

Commit:
Change hard-coded tolerances to libMesh::TOLERANCE

Commit:
Improve message in mooseAssert for BSpline

GiudGiud committed:
Address review on BSpline base util
- remove docstring from source file
- add more asserts on routine inputs
- reorder header includes
- remove reference to trivially copyable Real

GiudGiud reviewed (approved):
@roystgnr if you want to take another look at the first PR in the order

I mostly cleaned up and improved the commit history (and will deal with the conflicts on the later PRs, oh well)

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3645135/), step Results summary on 7412ddd wanted to post the following:

## Framework test summary

Compared against aec0a32 in job [civet.inl.gov/job/3640883](https://civet.inl.gov/job/3640883).

### No added tests


### Run time changes

| Test                                                            | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|-----------------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `multiapps/picard.iteration_adaptive`                           | 3.96       | 6.70       | +69.41% | 151.86      | 149.72      |
| `multiapps/secant.variables_transient/test_rerun`               | 2.42       | 3.83       | +58.03% | 115.00      | 121.24      |
| `multiapps/steffensen.variables_transient/test_rerun`           | 2.09       | 3.29       | +57.64% | 122.11      | 126.79      |
| `transfers/transfer_interpolation.test`                         | 2.46       | 3.81       | +54.74% | 156.92      | 143.28      |
| `executioners/multisys-ss-detection.multi-sys-ss`               | 4.99       | 7.66       | +53.55% | 136.29      | 139.71      |
| `convergence/reference_residual_convergence.zero_tolerance_ref` | 4.28       | 6.55       | +53.16% | 124.63      | 127.24      |
## Modules test summary

Compared against aec0a32 in job [civet.inl.gov/job/3640883](https://civet.inl.gov/job/3640883).

### No added tests


### Run time changes

| Test                                                                    | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|-------------------------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `navier_stokes/test:finite_element/ins/boussinesq.threaded_exo`         | 2.68       | 4.35       | +62.62% | 0.00        | 164.84      |
| `solid_mechanics/test:lagrangian/cartesian/total/rates.truesdell_shear` | 2.49       | 3.92       | +57.06% | 125.75      | 145.09      |
| `solid_mechanics/test:combined_creep_plasticity.stress_prescribed`      | 3.79       | 5.72       | +51.08% | 126.89      | 122.51      |

GiudGiud committed:
Fix doubling in BSpline

GiudGiud committed:
Add more documentation / explanation on what we are doing with the knot vector

GiudGiud committed:
Add more explanation on the features of this spline

GiudGiud commented on framework/src/utils/BSpline.C:
```diff
@@ -47,7 +47,8 @@ BSpline::buildKnotVector() const
   std::vector<libMesh::Real> knot_vector(_degree, 0); // initialize bottom to zeros
-  for (const auto i : make_range(num_points_left))
+
+  for (const auto i : make_range(num_points_left + 1))
```
Added more. I think this is less general than I thought. We are creating a uniform clamped B-Spline. Both of these adjectives are important here. They are what I want for the project afaik, but less general

GiudGiud commented:
Thanks for the review! 

No need to look at PR 2 and 3 until I rebase them and remove any inconsistency
